onnxruntime/core/framework/execution_provider.cc

```diff
@@ -52,13 +52,70 @@ IExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
 #endif
 }
 
+// Returns true if an allocator was found and replaced
+// Optionally user can provide an OrtMemoryInfo to swap in place
+// of an OrtMemoryInfo whose corresponding allocator is being replaced.
+static bool FindAndReplaceAllocator(const OrtMemoryInfo& mem_info,
+                                    MemoryInfoSet& mem_info_set,
+                                    AllocatorMap& allocators,
+                                    AllocatorPtr replacing_allocator,
+                                    const OrtMemoryInfo* mem_info_to_be_swapped_in = nullptr) {
+  auto ite = mem_info_set.find(mem_info);
+  if (ite != mem_info_set.end()) {
+    const int key = MakeKey(mem_info.id, mem_info.mem_type);
+    allocators[key] = replacing_allocator;
+
+    if (mem_info_to_be_swapped_in != nullptr) {
+      mem_info_set.erase(ite);
+      mem_info_set.insert(*mem_info_to_be_swapped_in);
+    }
+    return true;
+  }
+
+  return false;
+}
+
 // Update allocator in the provider if already present; ignore if not.
 void IExecutionProvider::ReplaceAllocator(AllocatorPtr allocator) {
   const auto& info = allocator->Info();
-  auto ite = mem_info_set_.find(info);
-  if (ite != mem_info_set_.end()) {
-    const int key = MakeKey(info.id, info.mem_type);
-    allocators_[key] = allocator;
+
+  if (FindAndReplaceAllocator(info, mem_info_set_, allocators_, allocator)) {
+    // We found an allocator corresponding to the provided
+    // allocator's OrtmemoryInfo, we hence return
+    return;
+  }
+
+  else {
+    // If we can't find an allocator registered with the exact OrtMemoryInfo
+    // as that of the replacing allocator, we do a "loosened" check
+    // (i.e.) check if there is an allocator registered with OrtAllocatorType
+    // as OrtArenaAllocator because for external user provided allocator
+    // we only accept OrtAllocatorType as OrtDeviceAllocator.
+    // If we do find such a registered allocator, we can safely go ahead
+    // and replace that with the provided allocator. This may seem like
+    // we are replacing an arena allocator with a non-arena allocator
+    // but in reality any user provided allocator may still be an arena
+    // allocator. We don't allow users to use OrtAllocatorType as
+    // OrtArenaAllocator for their allocators because we reserve its usage
+    // for our internal BFCArena.
+    // TODO: Should we remove the OrtAllocatorType field from OrtmemoryArena to
+    // avoid such problems and also remove the unintuitive phenomenon of binding
+    // the allocator type info to OrtmemoryInfo (which loosely is just device info) ?
+    const auto& original_info = allocator->Info();
+
+    // If the alloc_type was OrtArenaAllocator already, then it is a no-op
+    if (original_info.alloc_type == OrtAllocatorType::OrtArenaAllocator) {
+      return;
+    }
+
+    auto check_info = original_info;
+
+    // Mutate the alloc_type
+    check_info.alloc_type = OrtAllocatorType::OrtArenaAllocator;
```

Comment:
Can we check for both mem infos inside FindAndReplace instead of this?

Reply:
Fixed as discussed offline